add initial health checker

fmuyassarov opened the pull request:
- move out http pkg from instrumentation pkg to its own
- Add a simple health-checking functionality that reports healthiness of the nri-rm based on the last configuration update result
- Add simple test where we supply syntactically incorrect configMap which should be detected by the health checker and should cause the pod readiness probe failure

Notes: @klihub I've tested this code with DaemonSet and in case of nri-rm failure it is easily noticeable by the user of the k8s that there is some issue with the nri-resmgr pod, and if the users lists the events of the pod, they can see exact log message that config component has failed for example due to incorrect field in the supplied configuration. And of course, the pod status is also another way of quickly detecting an issue with the pod, because it will enter into CrashBackLoopOff status immediately after passing certain threshold of liveness probe timeout.

exxample: 
```bash
$ kubectl describe pod -n kube-system  nri-resmgr-kj6f5 
```
```sh
  Type     Reason     Age                    From               Message
  ----     ------     ----                   ----               -------
  Normal   Scheduled  11m                    default-scheduler  Successfully assigned kube-system/nri-resmgr-kj6f5 to n4c16-generic-ubuntu2204-containerd
  Normal   Started    7m37s (x3 over 11m)    kubelet            Started container nri-resmgr-topology-aware
  Warning  Unhealthy  7m28s (x9 over 7m52s)  kubelet            Liveness probe failed: HTTP probe failed with statuscode: 500
  Normal   Killing    7m28s (x3 over 7m46s)  kubelet            Container nri-resmgr-topology-aware failed liveness probe, will be restarted
  Warning  BackOff    7m28s (x2 over 7m28s)  kubelet            Back-off restarting failed container nri-resmgr-topology-aware in pod nri-resmgr-kj6f5_kube-system(0ee8c450-85ae-4337-8b22-ca407b030460)
  Normal   Created    7m13s (x4 over 11m)    kubelet            Created container nri-resmgr-topology-aware
  Normal   Pulled     6m13s (x6 over 11m)    kubelet            Container image "nri-resmgr-topology-aware:unknown" already present on machine
```

This is an initial code base and currently I'm working on adding another health checker for the NRT CRD. In other words, as @klihub already mentioned, we will check if the NRT CR was updated (fortunately in v1alpha2 of the NRT API that we are using, there is a field to track the last updated timestamp) whenever there are some containers added/removed.

fmuyassarov commented:
@klihub @jukkar PTAL

fmuyassarov commented:
One thing to note is, I didn't added the liveness probes to the balloons deployment file and didn't enable the instrumentation in the configuration files. Currently, it is only enabled and added to TA plugin tests. If you think, we need to add it for the balloons, please let me know.

jukkar commented on cmd/topology-aware/nri-resmgr-topology-aware-deployment.yaml.in:
```diff
@@ -87,6 +87,13 @@ spec:
               fieldRef:
                 fieldPath: spec.nodeName
           image: IMAGE_PLACEHOLDER
+          livenessProbe:
+            httpGet:
+              path: /healthz
+              port: 8891
+            initialDelaySeconds: 3
+            periodSeconds: 3
+            failureThreshold: 3
```
Please put the changes to `cmd/topology-aware/nri-resmgr-topology-aware-deployment.yaml.in` to a different commit than the `e2e: add basic health-....` as that should only have stuff for this test case.

fmuyassarov replied:
yes sure

jukkar commented on test/e2e/policies.test-suite/topology-aware/n4c16/test00-basic-placement/nri-resmgr.cfg.in:
```diff
@@ -5,6 +5,8 @@ policy:
   topology-aware:
     ColocatePods: $(echo ${COLOCATE_PODS:-false})
     ColocateNamespaces: $(echo ${COLOCATE_NAMESPACES:-false})
+instrumentation:
+  HTTPEndpoint: ":8891"
```
Do we really need add these
```
instrumentation:
  HTTPEndpoint: ":8891"
```
lines to all *.in files as those tests do not really need them (we do not have any support for liveliness in those tests).

fmuyassarov replied:
the reason I am adding them here is that we are using the same daemonSet for all the TA tests and I've added the liveness probe there. If we don't enable instrumentation in those TA tests then liveness probe fail so the test too.

klihub replied:
I agree, I think we should have a dedicated test case for the liveness probe / health check which then will need the endpoint enabled, but only for that test case.

fmuyassarov replied:
@jukkar @klihub sorry for the confusion. Let me clarify it a bit why I had to add `HTTPEndpoint` in other TA tests config files.

We have a single [nri-resmgr-topology-aware-deployment.yaml.in](https://github.com/jukkar/nri-resmgr/blob/main/test/e2e/files/nri-resmgr-topology-aware-deployment.yaml.in) that is used for all the TA plugin tests. We don't have/generate unique deployment files (daemonSet object) for each test case even though we have different configuration file for them. It is true that e2e is structured in a way that if you pass a custom configuration file to the test case under the directory `testxx-abc/` then that particular test will use that instead of default one (test/e2e/policies.test-suite/topology-aware/nri-resmgr.cfg).

I have added a separate test case for the health-checker `test12-health-checking`, which requires enabling HTTP endpoint on the configuration file it is using + readiness probe on the deployment file. This test case can have it is own configuration file but it is not needed. Because, the idea of the test is - let's launch the nri-rm with the default configuration file (doesn't matter if the HTTP endpoint is given or not) and then create a configMap with HTTP enabled + some syntactically incorrect data which should case the readiness probe to fail. 

If readiness probe is enabled on the daemonSet file but nri-rm is not exposing HTTP endpoint, then the container will fail (not considered as running). Since a single deployment file is used across all the tests of TA, and since I added readiness probe to the same file, I had to add HTTP endpoint to the custom configuration file of other tests and to the default one. Otherwise, we end up with a daemonSet with readiness probe enabled while nri-rm not exposing the HTTP endpoint and that will cause the failure for other tests.

I could add a separate daemonSet YAML for only health-checker test, but we don't usually create custom daemonSet and as said it is only configuration file that might change.

jukkar commented on test/e2e/policies.test-suite/topology-aware/n4c16/test12-health-checking/code.var.sh:
```diff
@@ -0,0 +1,25 @@
+vm-command "kubectl apply -f /etc/nri-resmgr/nri-resmgr-configmap.yaml"
+
+# Give enough time for the Kubernetes to consider the container unhealthy
+sleep 30
```
Can we make this shorter? Now the test pipeline is stalled for 30secs which is quite long.

fmuyassarov replied:
Yes, I will decrease it maybe to 20, because for the pod to fail it takes at least 15 sec because of the way I have configured the liveness probe the in deployment file.

klihub replied:
I think it should be possible to shorten that using short explicitly set `periodSeconds` and 
a `failureThreshold` of 1.

fmuyassarov replied:
yes indeed. Currently I have used 3 but I will decrease it. It was just random number I was using.

jukkar replied:
The shorter the better, this test can have as short value as needed.

jukkar commented on test/e2e/policies.test-suite/topology-aware/n4c16/test12-health-checking/code.var.sh:
```diff
@@ -0,0 +1,25 @@
+sleep 30
+
+# Check if we have CrashLoopBackOff status for the pod
+podstat="$(vm-command-q "kubectl get pod "$POD" -n kube-system | tail -1 | awk '{ print \$3 }'")"
```
Where is the `POD` variable coming from?
Couldn't we just call `vm-wait-pod-regexp` here?

fmuyassarov replied:
It is coming from https://github.com/jukkar/nri-resmgr/blob/b92ed0f23a5461effde8b8718e61a51b0da58051/test/e2e/run.sh#L246.

jukkar replied:
~Hmm, does this test really pass?~ I would have thought that the POD variable is not seen in this shell script.
My bad, I suppose the POD variable is now seen outside of launch() function, this was not meant to be like that as now we are depending on internal implementation of the launch() function.

jukkar commented on pkg/instrumentation/instrumentation.go:
```diff
@@ -17,7 +17,7 @@ package instrumentation
 import (
```
Minor nit to commit `pkg/http: move out....`
Please add to the commit body the reasoning why we are moving this, perhaps something like
`Moving the directory in order to allow it to be used in other parts of the nri-resmgr`
or something like that.

klihub commented:
> * move out http pkg from instrumentation pkg to its own
> * Add a simple health-checking functionality that reports healthiness of the nri-rm based on the last configuration update result
> * Add simple test where we supply syntactically incorrect configMap which should be detected by the health checker and should cause the pod liveness probe failure
> 
> Notes: @klihub I've tested this code with DaemonSet and in case of nri-rm failure it is easily noticeable by the user of the k8s that there is some issue with the nri-resmgr pod, and if the users lists the events of the pod, they can see exact log message that config component has failed for example due to incorrect field in the supplied configuration. And of course, the pod status is also another way of quickly detecting an issue with the pod, because it will enter into CrashBackLoopOff status immediately after passing certain threshold of liveness probe timeout.

But do we have any effect on the liveness probe timeout and indirectly when the pod will be restarted ? This was my only and major concern when we discussed this with our Polish colleagues who suggested using the health check mechanism to indicate configuration error.

It simply does not make sense/will not help to restart the pod. The config will be internally rolled back to the last/previous one that was successfully taken into use. So neither is a need to restart the pod nor will it accomplish anything good.

> 
> exxample:
> 
> ```shell
> $ kubectl describe pod -n kube-system  nri-resmgr-kj6f5 
> ```
> 
> ```shell
>   Type     Reason     Age                    From               Message
>   ----     ------     ----                   ----               -------
>   Normal   Scheduled  11m                    default-scheduler  Successfully assigned kube-system/nri-resmgr-kj6f5 to n4c16-generic-ubuntu2204-containerd
>   Normal   Started    7m37s (x3 over 11m)    kubelet            Started container nri-resmgr-topology-aware
>   Warning  Unhealthy  7m28s (x9 over 7m52s)  kubelet            Liveness probe failed: HTTP probe failed with statuscode: 500
>
>   Normal   Killing    7m28s (x3 over 7m46s)  kubelet            Container nri-resmgr-topology-aware failed liveness probe, will be restarted
Huston, we have a problem... with that "Normal Killing" an "will be restarted" above...
>
>   Warning  BackOff    7m28s (x2 over 7m28s)  kubelet            Back-off restarting failed container nri-resmgr-topology-aware in pod nri-resmgr-kj6f5_kube-system(0ee8c450-85ae-4337-8b22-ca407b030460)
>   Normal   Created    7m13s (x4 over 11m)    kubelet            Created container nri-resmgr-topology-aware
>   Normal   Pulled     6m13s (x6 over 11m)    kubelet            Container image "nri-resmgr-topology-aware:unknown" already present on machine
> ```
> 
> This is an initial code base and currently I'm working on adding another health checker for the NRT CRD. In other words, as @klihub already mentioned, we will check if the NRT CR was updated (fortunately in v1alpha2 of the NRT API that we are using, there is a field to track the last updated timestamp) whenever there are some containers added/removed.

Can't that be done by simply remembering the last error from the NRT CR update ? I mean, if it fails, we should get an error from it. We could simply store that and the registered health checer for the CRD update could simply use that to decide whether to report healthy or unhealthy with a logger error...

jukkar commented on pkg/agent/config-updater.go:
```diff
@@ -58,6 +59,8 @@ func newConfigUpdater(setConfig SetConfigFn) (configUpdater, error) {
 }
 
 func (u *updater) Start() error {
+	u.Info("Regsitering health-checker")
```
typo s/Regsitering/Registering/

jukkar commented:
> One thing to note is, I didn't added the liveness probes to the balloons deployment file and didn't enable the instrumentation in the configuration files. Currently, it is only enabled and added to TA plugin tests. If you think, we need to add it for the balloons, please let me know.

I think we need to support balloons from day 1, otherwise we just accumulate technical dept unnecessarily.

klihub commented:
> 
> But do we have any effect on the liveness probe timeout and indirectly when the pod will be restarted ? This was my only and major concern when we discussed this with our Polish colleagues who suggested using the health check mechanism to indicate configuration error.
> 
> It simply does not make sense/will not help to restart the pod. The config will be internally rolled back to the last/previous one that was successfully taken into use. So neither is a need to restart the pod nor will it accomplish anything good.

@fmuyassarov Can you check what happens, if you set `periodSeconds` to something suitable (maybe 15 or 30 seconds) and set `failureThreshold` so that the `periodSeconds * failureThreshold` is in hours, so maybe something like

```
  periodSeconds: 30
  failureThreshold: 240
```

If I understand how to [configure probes](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/#configure-probes) correctly, then this should have the effect that the pod goes into `NotReady` state (or at least not be in `Ready` state) but to not get restarted unless that conditions goes on for 2 hours (which is 7200 seconds = 240 * 30 seconds)...

klihub commented:
> Can't that be done by simply remembering the last error from the NRT CR update ? I mean, if it fails, we should get an error from it. We could simply store that and the registered health checker for the CRD update could simply use that to decide whether to report healthy or unhealthy with a logger error...

@fmuyassarov I think it would be more important to make sure we can go with this solution on the CRI-RM side by backporting it there. But that requires figuring out a workaround to prevent the pod from being so aggressively restarted by the control plane. So I'd prefer testing the `failureThreshold` workaround mentioned to see if we can go with that, then cleaning up the patch set according to Jukka's instructions, then merging it, and then doing the NRT CRD health check in a subsequent PR.

@jukkar @fmuyassarov WDYT ?

fmuyassarov commented:
> But that requires figuring out a workaround to prevent the pod from being s

@klihub @jukkar 
I have now tried with the ReadnessProbe instead of LivenessProbe and I can see that ReadnessProbe doesn't restart the container/pod and it just waits until the issue is resolved. The another diff is that, with ReadnessProbe pod is not intering the CrashBackLoopOff state but it just stays Running however, a ready container shows 0, because it is broken. Do you think this would be ok to use ReadnessProbe then?

fmuyassarov commented:
> > But that requires figuring out a workaround to prevent the pod from being s
> 
> @klihub @jukkar I have now tried with the ReadnessProbe instead of LivenessProbe and I can see that ReadnessProbe doesn't restart the container/pod and it just waits until the issue is resolved. The another diff is that, with ReadnessProbe pod is not intering the CrashBackLoopOff state but it just stays Running however, a ready container shows 0, because it is broken. Do you think this would be ok to use ReadnessProbe then?

something like this
```sh
vagrant@n4c16-generic-ubuntu2204-containerd:~$  kubectl get pods -A
NAMESPACE     NAME                                                          READY   STATUS    RESTARTS   AGE
kube-system   cilium-operator-864b7b486c-28jw6                              1/1     Running   0          27m
kube-system   cilium-q2pf5                                                  1/1     Running   0          27m
kube-system   coredns-787d4945fb-5xxn6                                      1/1     Running   0          27m
kube-system   coredns-787d4945fb-zgvpd                                      1/1     Running   0          27m
kube-system   etcd-n4c16-generic-ubuntu2204-containerd                      1/1     Running   0          27m
kube-system   kube-apiserver-n4c16-generic-ubuntu2204-containerd            1/1     Running   0          27m
kube-system   kube-controller-manager-n4c16-generic-ubuntu2204-containerd   1/1     Running   0          27m
kube-system   kube-proxy-77ltc                                              1/1     Running   0          27m
kube-system   kube-scheduler-n4c16-generic-ubuntu2204-containerd            1/1     Running   0          27m
kube-system   nri-resmgr-q8nv2                                              0/1     Running   0          5m19s
```

```sh
pod logs

I0224 13:46:07.004786       1 watcher.go:110] pushing default configuration to client
I0224 13:46:07.004800       1 config-updater.go:73] scheduling update after 2s rate-limiting timeout...
I0224 13:46:09.005349       1 resource-manager.go:438] applying new configuration from agent...
I0224 13:46:09.006101       1 config.go:142] validating configuration...
E0224 13:46:09.006134       1 resource-manager.go:465] configuration rejected: config error: implicit module main: given configuration data foo
E0224 13:46:09.221650       1 log.go:478] component config reported unhealthy: resource-manager: configuration rejected: config error: implicit module main: given configuration data foo
E0224 13:46:12.220004       1 log.go:478] component config reported unhealthy: resource-manager: configuration rejected: config error: implicit module main: given configuration data foo
E0224 13:46:15.221433       1 log.go:478] component config reported unhealthy: resource-manager: configuration rejected: config error: implicit module main: given configuration data foo
E0224 13:46:15.224886       1 log.go:478] component config reported unhealthy: resource-manager: configuration rejected: config error: implicit module main: given configuration data foo

```

klihub commented:
> > But that requires figuring out a workaround to prevent the pod from being s
> 
> @klihub @jukkar I have now tried with the ReadnessProbe instead of LivenessProbe and I can see that ReadnessProbe doesn't restart the container/pod and it just waits until the issue is resolved. The another diff is that, with ReadnessProbe pod is not intering the CrashBackLoopOff state but it just stays Running however, a ready container shows 0, because it is broken. Do you think this would be ok to use ReadnessProbe then?

Yes, definitely. Much saner idea than fiddling around with timeouts. I failed to spot it in the documentation.

klihub commented on cmd/topology-aware/nri-resmgr-topology-aware-deployment.yaml.in:
```diff
@@ -136,6 +143,8 @@ data:
     logger:
       Debug: nri-resmgr,resource-manager,cache,policy
+    instrumentation:
+      HTTPEndpoint: ":8891"
```
This probably shouldn't be `instrumentation/HTTPEndpoint` any more but split out/up to reflect the package having been moved out from `instrumentation`.

fmuyassarov replied:
~Is it going to pass ?~

klihub commented on pkg/healthz/healthz.go:
```diff
@@ -0,0 +1,104 @@
+
+const (
+	// just an example, we need to figure out what/if granularity makes any sense
+	Healthy Status = iota
+	Degraded
+	NonFunctional
+	NuclearMeltdown
```
I think we could update these to a more final granularity already in the initial commit. At least I'd take out the `NuclearMeltdown`.

fmuyassarov replied:
ah good point. Let me clean that up.

fmuyassarov commented:
I have addressed I think most of the comments. Let me re-run all the test once more and then I will open it back again for the final review.

fmuyassarov commented:
It seems e2e tests are broken in general. I'm investigating the issue.

fmuyassarov commented:
@jukkar @klihub thanks for the reviews. I have addressed your comments except two:
1. https://github.com/jukkar/nri-resmgr/pull/60#discussion_r1118402638 for removing instrumentation from the config file for the endpoint field. I'm not sure how the structure is enforced but I moved out HTTPEndpoint to the top level and it didn't work, nri-rm didn't accept that type of configuration. But that's perhaps `http.Start` is part of instrumentation/service.go, which should be done inside the http/http.go I think. If you think that's the reason, I can do it in a separate patch perhaps.
2. I asked in https://github.com/jukkar/nri-resmgr/pull/60#issuecomment-1443614640 whether we should add the same health-checker test for the balloons and @jukkar suggested that it would be good to add. I have no objections, but does it make sense to replicate the same tests among different plugins? Because, it is kind of generic functionality and not related to any specific plugin/policy. That's why I hesitated to add the same test in balloons tests dir, but if you think it would be beneficial, I will do it.

fmuyassarov commented:
> @jukkar @klihub thanks for the reviews. I have addressed your comments except two:
> 
> 1. [add initial health checker #60 (comment)](https://github.com/jukkar/nri-resmgr/pull/60#discussion_r1118402638) for removing instrumentation from the config file for the endpoint field. I'm not sure how the structure is enforced but I moved out HTTPEndpoint to the top level and it didn't work, nri-rm didn't accept that type of configuration. But that's perhaps `http.Start` is part of instrumentation/service.go, which should be done inside the http/http.go I think. If you think that's the reason, I can do it in a separate patch perhaps.

Or is it because HTTPEndpoint is defined inside the instrumentation/flags.go?

klihub commented:
> @jukkar @klihub thanks for the reviews. I have addressed your comments except two:
> 
> 1. [add initial health checker #60 (comment)](https://github.com/jukkar/nri-resmgr/pull/60#discussion_r1118402638) for removing instrumentation from the config file for the endpoint field. I'm not sure how the structure is enforced but I moved out HTTPEndpoint to the top level and it didn't work, nri-rm didn't accept that type of configuration. But that's perhaps `http.Start` is part of instrumentation/service.go, which should be done inside the http/http.go I think. If you think that's the reason, I can do it in a separate patch perhaps.

We can address that in a subsequent PR...

> 2. I asked in [add initial health checker #60 (comment)](https://github.com/jukkar/nri-resmgr/pull/60#issuecomment-1443614640) whether we should add the same health-checker test for the balloons and @jukkar suggested that it would be good to add. I have no objections, but does it make sense to replicate the same tests among different plugins? Because, it is kind of generic functionality and not related to any specific plugin/policy. That's why I hesitated to add the same test in balloons tests dir, but if you think it would be beneficial, I will do it.

@fmuyassarov 
It is common functionality, but there are two reasons why it makes sense to have test cases for both policies.
1. There is no separate agent any more. It is now part of the policy plugin binary.
2. Think about the data flow, when we get syntactically correct configuration data which is only semantically incorrect for the the policy. The policy needs to correctly indicate this (by returning an error) which gets propagated back to the agent and then via the readiness probe all the way to the control plane. That's a long chain which requires correct behavior at the head of the chain. It is safer to simply have a test for it (per policy).

fmuyassarov commented:
> * any more. It is now part of the policy plugin binary.
> * Think about the data flow, when we get syntactically correct configuration data which is only semantically incorrect for the the policy. The policy needs

Got it. I will add it then to the balloons policy too. Thanks for the thoughts.

jukkar commented on test/e2e/policies.test-suite/balloons/n4c16/test10-health-checking/code.var.sh:
```diff
@@ -0,0 +1,19 @@
+# Test that
+# Syntatically incorrect configuration file is resulting in the nri-resmgr
+# pod readiness probe failure.
+host-command "$SCP \"${SRC_DIR}/test/e2e/policies.test-suite/balloons/n4c16/test10-health-checking/nri-resmgr-configmap.yaml\" $VM_HOSTNAME:" || {
+    command-error "copying nri-resmgr-configmap.yaml to VM failed"
+}
```
This should be replaced by `vm-put-file` as it hides the paths and other nastiness from the user.
See for example `policies.test-suite/balloons/n4c16/test06-update-configmap/code.var.sh` for usage example.

jukkar commented on test/e2e/policies.test-suite/balloons/n4c16/test10-health-checking/nri-resmgr-configmap.yaml:
```diff
@@ -0,0 +1,26 @@
+    Klog:
+      skip_headers: true
+  foo: |+
+    bar: "abc"
```
Newline missing at the end of this file.

jukkar commented on test/e2e/policies.test-suite/topology-aware/n4c16/test12-health-checking/code.var.sh:
```diff
@@ -0,0 +1,19 @@
+# Test that
+# Syntatically incorrect configuration file is resulting in the nri-resmgr
+# pod readiness probe failure.
+host-command "$SCP \"${SRC_DIR}/test/e2e/policies.test-suite/topology-aware/n4c16/test12-health-checking/nri-resmgr-configmap.yaml\" $VM_HOSTNAME:" || {
+    command-error "copying nri-resmgr-configmap.yaml to VM failed"
+}
```
ditto

jukkar commented on test/e2e/policies.test-suite/balloons/n4c16/test10-health-checking/code.var.sh:
```diff
@@ -0,0 +1,17 @@
+# Test that
+# Syntatically incorrect configuration file is resulting in the nri-resmgr
+# pod readiness probe failure.
+vm-put-file "${SRC_DIR}/test/e2e/policies.test-suite/balloons/n4c16/test10-health-checking/nri-resmgr-configmap.yaml" nri-resmgr-configmap.yaml
```
We should avoid putting the full path here, it is just a pain later if changes are needed.
I think you can say here `vm-put-file $(instantiate nri-resmgr-configmap.yaml) nri-resmgr-configmap.yaml` the instantiate func should automatically figure out where the source file is.

fmuyassarov replied:
@jukkar  it doesn't seem to work
```sh
vm-put-file $(instantiate nri-resmgr-configmap.yaml) nri-resmgr-configmap.yaml
```

```shell
error: error instantiating "nri-resmgr-configmap.yaml": missing template 
host> scp -F /home/fmuyassarov/go/src/nri-resmgr/test/e2e/n4c16-generic-fedora37-containerd/.ssh-config "nri-resmgr-configmap.yaml" n4c16-generic-fedora37-containerd:"vm-put-file.nri-resmgr-configmap.yaml"
nri-resmgr-configmap.yaml: No such file or directory
command:     scp -F /home/fmuyassarov/go/src/nri-resmgr/test/e2e/n4c16-generic-fedora37-containerd/.ssh-config "nri-resmgr-configmap.yaml" n4c16-generic-fedora37-containerd:"vm-put-file.nri-resmgr-configmap.yaml"
output:      nri-resmgr-configmap.yaml: No such file or directory
exit status: 1
error:       failed to copy file to VM
host> scp -F /home/fmuyassarov/go/src/nri-resmgr/test/e2e/n4c16-generic-fedora37-containerd/.ssh-config n4c16-generic-fedora37-containerd:nri-resmgr.output.txt "/home/fmuyassarov/go/src/nri-resmgr/test/e2e/n4c16-generic-fedora37-containerd/policies.test-suite/balloons/test10-health-checking/"
scp: nri-resmgr.output.txt: No such file or directory
copying "-resmgr.output.txt" from VM failed
```

jukkar replied:
try by renaming the `nri-resmgr-configmap.yaml` to `nri-resmgr-configmap.yaml.in` in the file system, but do not add the .in in the script. The instantiate function will search the .in file and do any replacements (there are none in this case) and then sets the filename properly with path.

fmuyassarov replied:
that helped, thanks

jukkar commented on test/e2e/policies.test-suite/topology-aware/n4c16/test12-health-checking/code.var.sh:
```diff
@@ -0,0 +1,17 @@
+# Test that
+# Syntatically incorrect configuration file is resulting in the nri-resmgr
+# pod readiness probe failure.
+vm-put-file "${SRC_DIR}/test/e2e/policies.test-suite/topology-aware/n4c16/test12-health-checking/nri-resmgr-configmap.yaml" nri-resmgr-configmap.yaml
```
same here

fmuyassarov commented:
@klihub PTAL.

klihub commented:
LGTM.
